### PR TITLE
Added `current-row` to `StandardTableView`

### LINKED ISSUE #2824: Implement row selection for `StandardTableView`
### Discussed in https://github.com/slint-ui/slint/discussions/2818

<div type='discussions-op-text'>

<sup>Originally posted by **christiansavels** June  3, 2023</sup>
Hello,

I'm discovering Slint 1.0.2 with Rust 1.69, I really like this gui but I have a constraint.
My application uses tables to select items.
I manage to build the table with StandardTableView but I can't retrieve the selected row. 
Could you give me a clue or should I consider generating a pseudo table with a grid?

I feed the table with the 'place_searched_users' method (data from database): 

```rust
fn place_searched_users(ui: App)
{
    let ui_thread = ui.as_weak();
    thread::spawn(move || { 
        let ui = ui_thread.clone();
        let do_it = slint::invoke_from_event_loop(move || {
            let ui = ui.unwrap();
            let searched: String = ui.get_searched_user().to_string();
            let users = get_users( searched); // Vec<Person>
            let row_data: Rc<VecModel<slint::ModelRc<StandardListViewItem>>> = Rc::new(VecModel::default());

           // for u in users {
            for (index, u) in users.iter().enumerate() {
                let items = Rc::new(VecModel::default());
                let lastname = StandardListViewItem::from(slint::format!("{}", u.lastname));
                let firstname = StandardListViewItem::from(slint::format!("{}", u.firstname));
                let mail = StandardListViewItem::from(slint::format!("{}", u.email));
                let city = StandardListViewItem::from(slint::format!("{} {}", u.address.zip, u.address.city));

                items.push(lastname.clone().into());
                items.push(firstname.clone().into());
                items.push(mail.clone().into());
                items.push(city.clone().into());

                row_data.push(items.clone().into());
            }

            ui.global::<TableViewPageAdapter>().set_row_data(row_data.clone().into());
        });

        match do_it {
            Ok(_) => {
                println!("Done...");
            },
            Err(e) => {
                eprintln!("{:?}", e);
            }
        };
    });
}
```

```rust
pub struct Person {
   #[serde(rename = "_id", skip_serializing_if = "Option::is_none")]
   pub id: Option<bson::oid::ObjectId>,
   pub firstname: String,
   pub lastname: String,                                                                                                  
   pub email: String,
   pub password: String,
   pub address: Address,
}
```

The slint file for the table : 

```
import { GroupBox, AboutSlint, StandardTableView, HorizontalBox, VerticalBox, Button, LineEdit  } from "std-widgets.slint";

export component UsersBox inherits StandardTableView {    
    vertical-stretch: 0;

    sort-ascending(index) => {
        TableViewPageAdapter.sort_ascending(index);
    }

    sort-descending(index) => {
        TableViewPageAdapter.sort-descending(index);
    } 
    
    columns: [
        { title: "Nom" },
        { title: "Prénom" },
        { title: "Mail" },
        { title: "Ville" },
    ];

    rows: TableViewPageAdapter.row_data;
}

export global TableViewPageAdapter {
    
    callback sort_ascending(int);
    callback sort_descending(int);
    
    in-out property <[[StandardListViewItem]]> row_data: [
        [ { text: "Item 1.1", clicked : {} }, { text: "Item 1.2" }, { text: "Item 1.3" }, { text: "Item 1.4" }, ],
        [ { text: "Item 2.1" }, { text: "Item 2.2" }, { text: "Item 2.3" }, { text: "Item 2.4" } ],
        [ { text: "Item 3.1" }, { text: "Item 3.2" }, { text: "Item 3.3" }, { text: "Item 3.4" }, ],
        [ { text: "Item 4.1" }, { text: "Item 4.2" }, { text: "Item 4.3" }, { text: "Item 4.4" }, ],
        [ { text: "Item 5.1" }, { text: "Item 5.2" }, { text: "Item 5.3" }, { text: "Item 5.4" }, ],
        [ { text: "Item 6.1" }, { text: "Item 6.2" }, { text: "Item 6.3" }, { text: "Item 6.4" }, ],
    ];
}
```

Thank you in advance.</div>

### PR DESCRIPTION
**Pull request by FloVanGH**
Fixes #2824

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -27,6 +27,8 @@ All notable changes to this project are documented in this file.
  - Added functions on color: `transparentize`, `mix`, and `with-alpha`.
  - Added a `close()` function and a `close-on-click` boolean property to `PopupWindow`.
  - Added basic translation infrastructure with `@tr("...")`
+ - Fixed `LineEdit`'s placeholder text not being rendered with the same font attributes as regular `LineEdit` text.
```
Comment (ogoffart) [inline CHANGELOG.md:30]:
```suggestion
```

I moved that line up in another commit

File: docs/language/src/builtins/widgets.md [hunk 2]
```diff
@@ -437,6 +437,7 @@ Same as [`ListView`](#listview), and in addition:
 -   **`current-sort-column`** (_out_ _int_): Indicates the sorted column. -1 mean no column is sorted.
 -   **`columns`** (_in-out_ _\[[`TableColumn`](structs.md#tablecolumn)\]_): Defines the model of the table columns.
 -   **`rows`** (_\[\[[`StandardListViewItem`](structs.md#standardlistviewitem)\]\]_): Defines the model of table rows.
+-   **`current-row`** (_in-out_ _int_): The index of the currently active row. -1 mean none is selected, which is the default.
```
Comment (ogoffart) [inline docs/language/src/builtins/widgets.md:440]: Should we also not document the set-current-row function ?

**Comment by tronical** [comment]
I think `current-row` makes sense if there's also a `current-column`, which there isn't.

In the long run I think we want to have the concept of selections - that can be a single cell, a row, a column, or a grid. When introducing that, we need to avoid the API becoming awkward.

That makes me lean towards the more "vague" `current-item`, because it's not evident that `current-row` covers the entire row or just specifies the row number of the current cell.

Either way this is an imperfect API, so the docs are key :)

**Comment by ogoffart** [comment]
> That makes me lean towards the more "vague" current-item, because it's not evident that current-row covers the entire row or just specifies the row number of the current cell.

When we have a way to select cell, current-row will be that that covers the entire row, so that would be a good name.
This leans toward choosing `current-row`

The current API of StandardTableView has a `rows` property for the content instead of a `model` in StandardListView, so it might make sense to have `current-row` so i tend to agree with the current naming of `current-row`

Maybe we could also deprecate `rows` and rename it `model` like StandardListView, but that's a bit out of scope of this patch and i'm tempted to prefer `current-row` now.

File: docs/language/src/builtins/widgets.md (hunk 2 above)
Comment (FloVanGH) [reply docs/language/src/builtins/widgets.md:440]: Yes I will add the documentation.

**Comment by tronical** [comment]
> > That makes me lean towards the more "vague" current-item, because it's not evident that current-row covers the entire row or just specifies the row number of the current cell.
> 
> When we have a way to select cell, current-row will be that that covers the entire row, so that would be a good name. This leans toward choosing `current-row`

But then moment we do that, the behavior changes: Instead of referring to an entire row, it will refer to just the row of a cell.

I was a bit unsure about that - I would've thought that a property that specifies the entire currently selected row would be -1 if only a single cell is selected.

If that change of meaning is okay though, then `current-row` is indeed the better name.

> The current API of StandardTableView has a `rows` property for the content instead of a `model` in StandardListView, so it might make sense to have `current-row` so i tend to agree with the current naming of `current-row`
> 
> Maybe we could also deprecate `rows` and rename it `model` like StandardListView, but that's a bit out of scope of this patch and i'm tempted to prefer `current-row` now.

**Comment by FloVanGH** [comment]
Yeah I think is best to stick now with `current-row`.

**Review by ogoffart — APPROVED** [review]

File: docs/language/src/builtins/widgets.md [hunk 3]
```diff
@@ -438,12 +438,17 @@ Same as [`ListView`](#listview), and in addition:
...
 
 ### Callbacks
 
 -   **`sort-ascending(`_`int`_`)`**: Emitted if the model should be sorted by the given column in ascending order.
 -   **`sort-descending(`_`int`_`)`**: Emitted if the model should be sorted by the given column in descending order.
 
+### Functions
+
+-   **`set-current-row(_index: int_)`: Sets the current row and brings it into view
```
Comment (ogoffart) [inline docs/language/src/builtins/widgets.md:450]:
```suggestion
-   **`set-current-row(index: int)`: Sets the current row and brings it into view
```

btw, the one in StandardListView is also wrong:
https://slint-ui.com/releases/1.0.2/docs/slint/src/builtins/widgets#functions
